### ISSUE TITLE
World coordinates API question

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#1016]
## ❓ Questions and Help

After discussions on Thursday I'm picking up work on #885 and https://github.com/jni/napari/pull/6 from @jni and @GenevieveBuckley to add support for world coordinates to the viewer (see previous discussions in #763 too and a number of comments from @d-v-b across these issues). I think I am fairly close to having a follow-on PR that has many of the features we want but I want to ask two questions before moving forward. I'm making a new issue to keep the discussion easy to follow and self-contained.

### Question 1:
**Can each layer just have a single nD affine transform that goes from data to world coordinates?** Ignore the complexity of the 3D rotation / arbitrary slicing stuff for now. I have a plan to fail gracefully and maybe one day support that if needed. Also ignore things like grid mode and image pyramids. What I really want to know is if people want more than that, i.e. piecewise affine or deformable transforms etc. and want our system to be constructed at a level of generality to allow that. I basically cannot see how to integrate those options with vispy and think that we should limit ourselves to affine, but want to check with @napari/core-devs before pushing further. I also think that each layer should only have one affine transform since affine transforms can be easily composed with each other and just having to keep track of one seems much easier to me.

### Question 2:
If the answer to Question 1 is yes, **what should the transform related API on the base layer look like?** Right now the base layer accepts `scale` and `translate`. Should it also accept `affine` and then compose transforms with `scale` and `translate`? If so what should it return when someone wants to do a roundtrip with layer? Should it only accept `affine` and nothing else? Should it accept `scale`, `translate`, a `rotation` matrix (and a `sheer` matrix?). Something else entirely?

I like layer accepting `scale`, `translate`, `rotation` and `sheer` the best as those 4 elements can be composed into an affine matrix which can then be decomposed to get the same elements back, even in nD (see https://github.com/matthew-brett/transforms3d/blob/master/transforms3d/affines.py for example). This allows us to maintain one source of ground truth internally, the affine matrix, but allows users to create it in an easy way.

We could also let a user pass an `affine` matrix directly (n+1 x n+1), but that would be an exclusive choice with the above, i.e. you couldn't also pass `scale`.

This would now mean we have 5 transform related keywords in the `add_*` methods, all optional, but one mutually exclusive with the other 4.

What do @napari/core-devs @d-v-b think?

**Comment by jni** [comment]
@sofroniewn 

> Can each layer just have a single nD affine transform that goes from data to world coordinates?

I don't love this. There's certainly a *massive* user base that uses more complex transforms, *and* vispy allows arbitrary transforms, *and* geodata *and*... In short, not accounting for more complex transforms cuts out a huge number of people. We can say "leave this for later", but honestly, is there a better time than now?

Part of the reason I chose to go with scale + translation as the only options in #885 is that it then maps very closely to what exists now. If we are adding the complexity of affine, I feel like it's just an extra step to go higher...

> what should the transform related API on the base layer look like?

I like your suggestions here. Incidentally for some reason I could never find an answer to the question of whether the affine matrix could be thus decomposed! Good job finding that. =)

Incidentally, this whole topic hides so many dragons. For example, when doing a z-stack where there is microscope drift over the acquisition, you need a different translation setting per plane. I suspect that allowing for such situations will look similar to allowing for the more complex transforms described above.

**Comment by d-v-b** [comment]
Yay! Thanks for working on this!

> Can each layer just have a single nD affine transform that goes from data to world coordinates?

For me, affine is enough (for now)! Neuroimaging formats and bigdataviewer describe data with a full affine transform. Just make sure that the clipping plane/volume is also represented by an affine transformable-object. Right now there is no clipping plane/volume. That functionality is localized in the `dims` object but it is not a geometrical object. `dims` should probably be upgraded to support spatial transforms, even if getting slices of data still uses array indexing under the hood.

For the future, it would be very cool if napari could support stuff like [BigWarp](https://github.com/saalfeldlab/bigwarp), but I wouldn't consider this core.

> If the answer to Question 1 is yes, what should the transform related API on the base layer look like?

Personally I'm OK with being forced to supply a full affine matrix XOR some kwargs. In both cases it would be nice to access e.g. `layer.scale` (just get the scale out) and `layer.affine` (get the full affine transform) as properties.

**Comment by sofroniewn** [comment]
@jni 
> >Can each layer just have a single nD affine transform that goes from data to world coordinates?
>
> I don't love this. There's certainly a massive user base that uses more complex transforms, and vispy allows arbitrary transforms, and geodata and... In short, not accounting for more complex transforms cuts out a huge number of people. We can say "leave this for later", but honestly, is there a better time than now?

Now is a good time to talk about it :-) I just have a hard time understanding how it would work in practice without substantial PRs to vispy too. While the vispy [BaseTransform](https://github.com/vispy/vispy/blob/master/vispy/visuals/transforms/base_transform.py#L25) is architected in a generic way it needs to be subclassed to be used and someone needs to provide both a python `map` function and a shader snippet `glsl_map` function along with inverses for both too - see the [STTransform](https://github.com/vispy/vispy/blob/master/vispy/visuals/transforms/linear.py#L54) for example. The only [pre-built transforms in their docs](http://vispy.org/visuals.html#module-vispy.visuals.transforms) are `Null`, `ST` (`scale` and `translate`), `Matrix` (same as `affine`), and `Log`. I don't think `Log` is so important that we have to support it too, which basically just leaves us with being able to use `affine` which is why I can imagine how to make our stuff use affine but not more :-)

I see how your work in #885 starts with a base Transform that allows for a generic callable function which is fine, but it's unclear how we could ever get that callable function into shader code that vispy would run without us providing it, and it is not clear to me that we want to role our own cpu version of those transforms.

Even if we had the shader code figure out, I also don't really have a clear idea of what the API for the layer would then look like, maybe we'd add `transform_func` and `transform_inv_func` and they would then become mutually exclusive with all the other keyword args, `scale`, `translate`, `rotate`, `sheer`, and `affine`? Would we be ok with these things being mutually exclusive and a layer just having a single transform, or would you want to be able to use these things together and have things chain? Chaining seems much more overhead to me. If we go the mutually exclusive single transform route then I guess `scale` etc should just return `None` like you do in #885 and we could handle that outside the layer, but it might be a bit strange / confusing. I could also see people wanting to set scales, and then apply deformable transforms. Thoughts? Should we go with a `TransformChain` for each layer or just a `Transform` for each layer?

Thanks @d-v-b for the link to Big Warp. Do you know if there has been any Big Warp stuff in Python? That could really help me reason about APIs. Maybe @bogovicj can let us know? And feel free to weigh in on any of the rest of the discussion too @bogovicj.

Also @jni any links to the geospatial transforms, examples, APIs etc would be helpful too.

> Part of the reason I chose to go with scale + translation as the only options in #885 is that it then maps very closely to what exists now. If we are adding the complexity of affine, I feel like it's just an extra step to go higher...

Ok, that's fair enough. I think I can push a scale + translation approach over the finish line based on the work you and @GenevieveBuckley have done. I'll focus on that as it won't change our API, but will at least improve things. The move to affine afterwards shouldn't actually be too hard and I know that's what @d-v-b wants!! We can then think about multiple transforms per layer and non-affine too

**Comment by GenevieveBuckley** [comment]
> I could also see people wanting to set scales, and then apply deformable transforms. Thoughts? Should we go with a TransformChain for each layer or just a Transform for each layer?

I can see people wanting to do this too. 

Isn't that why we wanted the chained transforms in the first place? Right now chained transforms just looks like overhead because we're only reproducing the current functionality available in napari, but it allows more flexibility later on. I suspect also that chaining transforms will make the pyramid and grid view modes a little less confusing to work with, since their effect will be cleanly separated into their own transform.

Note: I hope it's clear I'm not actually arguing against any of your ideas @sofroniewn, just trying to get a little clearer about this in my own head. Thanks so much for diving in, you really do a ton to make sure this project keeps momentum :smile:

**Comment by jni** [comment]
> I don't think Log is so important that we have to support it too, which basically just leaves us with being able to use affine which is why I can imagine how to make our stuff use affine but not more :-)

But that's why there are packages like [vismap](https://github.com/codypiersall/vismap) which provide additional transforms — which we probably want to support! And other users (*cough* @bogovicj *cough*) will probably be able to directly add support for additional transforms. I don't know for sure but I do think this is one of the easier-to-extend parts of vispy, since transforms are conceptually very simple: take a vec3 and return a vec3.

> I could also see people wanting to set scales, and then apply deformable transforms. Thoughts? Should we go with a TransformChain for each layer or just a Transform for each layer?

Yeah, what @GenevieveBuckley said! The TransformChain makes it much clearer what the transforms are doing in relation to grid mode and pyramids. I'll point out that my proposal to use a chain in the first place came from chats with @axtimwalde about what BDV is doing. That is one package where I'm pretty sure the implementation is well thought through. :joy: 

In short: chain, chain, chain! Any slowness can be relieved by async rendering of blocks, which we'll have to do soon anyway.

**Comment by bogovicj** [comment]
Thanks for looping me in everyone! 

> There's certainly a massive user base that uses more complex transforms

:+1: My use case(s) require more complicated transformations that affine, and almost always involves chaining a sequence of transforms.

> Do you know if there has been any Big Warp stuff in Python?

I've seen stuff that builds transforms from point-correspondences, but havn't seen something that's nicely interactive.

> (cough @bogovicj cough) will probably be able to directly add support for additional transforms

lol.  Happy to! :smiley:  #subtlety 

## thoughts on API

* For general transforms, it's nice to have an api that differentiates between invertible and non-(easily)-invertible transforms.
* It'll take some work, but will be nice to avoid the overhead caused by chains of transforms if the code/developer knows how to simplify them.  E.g. a chain consisting of entirely affines can be simplified to a single affine. Whether this is the dev's job or napari's job is a decision to make.  I'd go with deferring to dev's for now, and putting this off to the future.

* Consider at some point differentiating transformations that are continuous (more general + slower & need interpolation) vs discrete (faster, no interpolation). E.g. "just" permuting axes is discrete, and might be worth having as part of a chain.  

### Inverses and api choices
It's good to be super clear about how transforms are applied, since it *can* get counter intuitive.  Specifically, most use cases may want something like:
* `transform(image, translation(x=-5.5))`, which we'd all expect moves the image 5.5 units "to the left"
   * This requires that the input transform be invertible.  Since it will use the inverse of the passed transform to do the rendering in order to get the expected result
   * (because one needs the target space - to moving space transform to render images with interpolation)

but some use cases will need the more general:
* `general_transform(image, some_nonlinear_transform) # needs a decent name`
   * where it's understood that `some_nonlinear_transform` transofrms target-to-moving space.

Edit: I notice that vispy's `BaseTransform` is rather [wishy-washy](https://github.com/vispy/vispy/blob/master/vispy/visuals/transforms/base_transform.py#L37-L42)
>    Optionally, an inverse() method returns a new transform performing the
    inverse mapping.
    Note that although all classes should define both map() and imap(), it
    is not necessarily the case that imap(map(x)) == x; there may be instances
    where the inverse mapping is ambiguous or otherwise meaningless.

I'm not crazy about that choice.

**Comment by sofroniewn** [comment]
Ok, I'm taking away from this that the group is very pro-chain transform and likes thinking about transforms more generally than affine :-) That's good to know and fine with me. I have made a super simple refactor in #1018 that changes no functionality, but now lets us use a `scale` / `translate` transform. If that looks good I'll add a "world" coordinate system in a followup, and then explore affine, and then explore more general. Along the way I'll think about chains and where different transforms should live.

@bogovicj thanks so much for weighing in!! As we get closer to some of these more advanced topics I will likely ping you again, it will take me a bit of time to digest your feedback on this issue.

**Comment by GenevieveBuckley** [comment]
> It'll take some work, but will be nice to avoid the overhead caused by chains of transforms if the code/developer knows how to simplify them. E.g. a chain consisting of entirely affines can be simplified to a single affine. Whether this is the dev's job or napari's job is a decision to make. I'd go with deferring to dev's for now, and putting this off to the future.

So vispy has a `simplified` method (see [here](https://github.com/vispy/vispy/blob/19f024229f5bebd32c6fb57d079b498848b65d16/vispy/visuals/transforms/chain.py#L99) and [here](https://github.com/vispy/vispy/blob/19f024229f5bebd32c6fb57d079b498848b65d16/vispy/visuals/transforms/chain.py#L252)) for chain transforms, which we'll absolutely want to use. 

There are some cases in which it won't completely collapse to a single matrix, eg: if some transforms in the chain are marked as "dynamic" - see https://github.com/vispy/vispy/issues/1249

**Comment by VolkerH** [comment]
>  It'll take some work, but will be nice to avoid the overhead caused by chains of transforms if the
> code/developer knows how to simplify them.  E.g. a chain consisting of entirely affines can be 
> simplified to a single affine. Whether this is the dev's job or napari's job is a decision to make.  I'd
> go with deferring to dev's for now, and putting this off to the future.

There is an important reason why you may not wan't to simplify a chain of affines into a single affine. I learned this the hard way (but tracking down the source of artefacts) and have communicated this in person to @jni a few times:

If you are dealing with unisotropic data, e.g. from confocal microscopes, where the voxel spacing in z is significantly larger than in x,y directions, you don't want to combine the scaling transform that makes the pixels isotropic with any of the other transforms (such as rotations) that may mix dimensions upon interpolation. The reason is that most implementations of affine transforms on GPUs (on CPUs too, I believe ITK is a notable exception) assume that distances should be equally weighted along any direction when performing interpolation. So a voxel that may be 100 um away (because it is along the z axis) will be given equal weight to a voxel that is only 5um (away) in x,y for most implementations. Not combining the transforms can avoid that if the first affine transform makes the data isotropic. This usually means upsampling and thus the storage and subsequent computational cost for the other transforms will go up.

There is no easy way around that at the moment, unless someone implements a nice OpenCL-based affine transform that takes non-isotropic voxels into account.

**Comment by bogovicj** [comment]
@VolkerH ,

I generally agree with everything you wrote, but am confused about this part:

>  Not combining the transforms can avoid that if the first affine transform makes the data isotropic. 

What's the problem of combining transforms here?  Is this what you have in mind:
1) Create an affine that makes the image isotropic.  Render that isotropic (upsampled) image into memory.
2) Apply your affine (in physical coordinates) on that upsampled image.

Why not do:
1) Create an affine that makes the image isotropic. and combine (compose) it with the affine in physical coordinates.
2) Use that combined affine with your favorite naive transformation library.

i.e. I don't see the downside of combining affines.  Actually, I think combining affines is *better* since there's only one interpolation step, and because you don't have to store the upsampled image.   (I'm thinking about tri-linear interpolation here)

Am I missing something?

**Comment by axtimwalde** [comment]
Here are my 5ct: I like to keep lists or chains of transformations around and reduce them as needed for the endpoint, i.e. when you want to render a frame.

The above mentioned hack to treat non-isotropic scaling independently from the rest of the transformation is a hack to avoid transforming the interpolation kernel correctly.  I find it less than exciting to bake in such hacks into a general purpose framework.  They are useful and should be used where needed but not dictate the design of the system.  The system should be designed to enable hacks, not to be a hack.

Also, in case you care, please feel free to inspect https://javadoc.scijava.org/ImgLib2/net/imglib2/view/Views.html and https://github.com/imglib/imglib2-realtransform for inspiration.  These are the ImgLib2 generic functional approaches to discrete and real-space transformations including a few quirky transformations that happened to be useful for various projects.

**Comment by jni** [comment]
@axtimwalde thank you so much for weighing in! And yes, we do care. =) I have a follow-up question:

> a hack to avoid transforming the interpolation kernel correctly.

To clarify, this means that the *interpolation software* (e.g. `scipy.ndimage.affine_transform`) needs to adjust its distance calculations to take the transform into account, is that right? ie in the case of a skew in x, the "nearest" points to a particular coordinate will be skewed, and this must be taken into account by the interpolation code, right? ie if this artifact is present, there is little to be done other than fix the software or use @VolkerH's hack of doing the transforms in sequence, right?

**Comment by VolkerH** [comment]
Sorry for not replying earlier, I overlooked this in the many Github notifications.

> Why not do:
> 
>     1. Create an affine that makes the image isotropic. and combine (compose) it with the affine in physical coordinates.
> 
>     2. Use that combined affine with your favorite naive transformation library.
> 
> 
> i.e. I don't see the downside of combining affines. Actually, I think combining affines is _better_ since there's only one interpolation step, and because you don't have to store the upsampled image. (I'm thinking about tri-linear interpolation here)

I had been thinking exactly the same. For one project, I combined several transformations that were done in sequence, thinking a single affine transformation (one interpolation step) should give better results than 3 affine transformations with interpolation steps each.  Then I noticed some funny artefacts with the single affine transform that did not occur when I did the transformations in sequence. (This was here https://github.com/VolkerH/Lattice_Lightsheet_Deskew_Deconv/issues/22 , although I was on the wrong track with my first attempts). 

I attempt to explain it. Assume you have an affine transformation A. Now you go through the destination image. A pixel in your destination image might map to coordinate  that lies
between integer pixel positions in the source image, say with an offset `delta r` and `delta c` as in the image I am pasting below. 

![image](https://user-images.githubusercontent.com/2210044/76033641-ca2b8300-5f90-11ea-8200-72a821b3ea6c.png)
(image from https://stackoverflow.com/questions/26142288/resize-an-image-with-bilinear-interpolation-without-imresize)

So you have 4 pixel values that you sum up, each with 2 weighting factors. If the voxel spacing in the real world is equal in both dimensions everything is fine. However, assume that the real-workd spacing in dimension 2 is different from those in  than in dimension 1.  Now, the values for `delta r` and `delta c` in the above image would have to be weighted accordingly.
If your first transformation only scales (upsamples) one dimension, the distance in the other dimension will always be zero (say `delta c`  for example here). So there is no mixing of pixel values. For any subsequent transformations it is then fine to weight voxels along each direction equally. 
Probably someone else can explain it more eloquently, but maybe it gives you an idea what's going on.

The `affine` functions in `vipsy`, `scipy.ndimage` or `gputools` don't allow passing in such weights. I looked into the OpenCL impementions some time back to see how easy it would be to pass in some additional weights for the dimensions. However, all the implementaions I found rely on some interpolators in the Khronos OpenCL code base that does not offer this option to pass in weights.
It should not be too hard to write a trilinear interpolation code that takes these weights into account using plain python or C, however integrating this into the existing OpenCL based GPU libraries, so that it works in vispy etc. is something that I don't have the skills or the time for currently.  

I assume that's what @axtimwalde 's code does (I haven't looked at the Imglib2 snippets) yet.

**Comment by VolkerH** [comment]
Just to clarify, sequencing the transforms is a hack and having a proper interpolator would be much nicer ... I just can't implement it at the level where it is needed to work in  vispy, napari, gputools.
For very anisotropic data one blows up the memory and compute requirements unecessarily by upscaling.

BTW: I am not 100% sure there is no support for additional weights in the Khronos libraries ... I am just saying I haven't seen anything obvious.

**Comment by GenevieveBuckley** [comment]
The level of detail in your comments is very helpful @VolkerH 😄

**Comment by sofroniewn** [comment]
I'm closing this as we now have support for chains of transforms and affines. Lots of great comments in this tread, and no doubt things we can still improve. Thanks all!!

**Comment by kevinyamauchi** [comment]
🎉 🎉 🎉
